core/shared/src/main/scala/com/lynbrookrobotics/potassium/streams/Stream.scala

```diff
@@ -0,0 +1,432 @@
+      ptr.get match {
+        case Some(s) =>
+          s.receiveB(b)
+        case None =>
+          bCancel()
+      }
+    }
+
+    ret
+  }
+
+  /**
+    * Zips a stream with the time of value emission, for use in situations such as
+    * synchronizing data sent over the network or calculating time deltas
+    *
+    * @return a stream with tuples of emitted values and the time of emission
+    */
+  def zipWithTime(implicit clock: Clock): Stream[(T, Time)] = {
+    map(v => (v, clock.currentTime))
+  }
+
+  /**
+    * Applies a fixed size sliding window over the stream
+    * @param size the size of the window
+    * @return a stream returning complete windows
+    */
+  def sliding(size: Int): Stream[Queue[T]] = {
+    var last = Queue.empty[T]
+
+    val ret = new Stream[Queue[T]] {
+      override val expectedPeriodicity = self.expectedPeriodicity
+    }
+
+    val ptr = WeakReference(ret)
+
+    var cancel: Cancel = null
+    cancel = this.foreach { v =>
+      ptr.get match {
+        case Some(s) =>
+          if (last.size == size) {
+            last = last.tail
+          }
+
+          last = last :+ v
+
+          if (last.size == size) {
+            s.publishValue(last)
+          }
+        case None =>
+          cancel()
+      }
+    }
+
+    ret
+  }
+
+  /**
+    * Applies an accumulator to the stream, combining the accumulated value
+    * with each emitted value of this stream to produce a new value
+    * @param initialValue the initial accumulated value
+    * @param f the accumulator to combine the value built and the latest emitted value
+    * @tparam U the type of the accumulated value
+    * @return a stream accumulating according to the given function
+    */
+  def scanLeft[U](initialValue: U)(f: (U, T) => U): Stream[U] = {
+    var latest = initialValue
+
+    map { v =>
+      latest = f(latest, v)
+      latest
+    }
+  }
+
+  /**
+    * Zips the stream with the periods between when values were published
+    * @return a stream of values and times in tuples
+    */
+  def zipWithDt(implicit clock: Clock): Stream[(T, Time)] = {
+    zipWithTime.sliding(2).map { q =>
+      (q.last._1, q.last._2 - q.head._2)
+    }
+  }
+
+  /**
+    * Calculates the derivative of the stream, producing units of
+    * the derivative of the stream's units
+    * @return a stream producing values that are the derivative of the stream
+    */
+  def derivative[D <: Quantity[D] with TimeDerivative[_]](implicit intEv: T => TimeIntegral[D], clock: Clock): Stream[D] = {
+    zipWithTime.sliding(2).map { q =>
+      val dt = q.last._2 - q.head._2
+      (q.last._1 / dt) - (q.head._1 / dt)
+    }
+  }
+
+  /**
+    * Calculates the integral of the signal, producing units of
+    * the integral of the signal's units
+    * @return a signal producing values that are the integral of the signal
+    */
+  def integral[I <: Quantity[I] with TimeIntegral[_]](implicit derivEv: T => TimeDerivative[I], clock: Clock): Stream[I] = {
```

Comment:
No no, that makes a lot of sense! Because `zipWithTime` ends up giving you the current time, instead of the time when the root element was emitted. I like this idea, a root time stream, which you can use when you have some calculation that is time dependent.